docs/design/proposals/direct-blk-device-assignment.md

```diff
@@ -0,0 +1,264 @@
+directory. The reason for adding a sandbox id file is to establish a mapping between the volume and the sandbox. 
+Later, when the kata-runtime handles the `get-stats` and `resize` commands, it can go to the volume’s directory, 
+find the sandboxes that mount this volume, and then identify the endpoint of the corresponding containerd-shim-kata-v2.
+
+### containerd-shim-kata-v2 changes
+containerd-shim-kata-v2 provides an API for sandbox management through a unix domain socket. Two new handlers are proposed:
+```
+New handlers:
+/fs-stats
+/fs-resize
+
+Example:
+
+curl —unix-socket <shim socket path> -I \
+-X GET ‘http://localhost/fs-stats' \
+-d ‘{ “block-device”: “/dev/sda” }’ 
+
+curl —unix-socket <shim socket path> -I \
+-X _PUT_ ‘http://localhost/fs-resize' \
+-d ‘{ “block-device”: “/dev/sda”,
+“volume-size”: “123123” 
+}’
+```
+
+The shim will then forward the corresponding request to the kata-agent to mount the filesystem, 
+gather the stats and refresh the filesystem metadata. For `fs-resize` operation, 
+kata-runtime might also need to notify the hypervisor to resize the block device (e.g. call `block_resize` in QEMU). 
+
+### Kata-agent changes
+
+For mounting a filesystem on the block device volume, the information can be passed to Kata-agent through the existing 
+`Storage` grpc object. Two new grpc request objects are added to the Kata-agent grpc interface for 
+resizing and getting volume stats:
+```Golang
+type ResizeFileSystemRequest struct {
+  mountPoint string
+  size       int64
+}
+
+type GetVolumeStatsRequest struct {
+  mountPoint string
+}
+```
+The Kata agent then handles the requests and returns the information back to the kata shim.
+
+### Step by step walkthrough
+
+Given the following definition:
+```YAML
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: app
+spec:
+  runtime-class: kata-qemu
+  containers:
+  - name: app
+    image: centos
+    command: ["/bin/sh"]
+    args: ["-c", "while true; do echo $(date -u) >> /data/out.txt; sleep 5; done"]
+    volumeMounts:
+    - name: persistent-storage
+      mountPath: /data
+  volumes:
+  - name: persistent-storage
+    persistentVolumeClaim:
+      claimName: ebs-claim
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  annotations:
+    skip-hostmount: "true"
+  name: ebs-claim
+spec:
+  accessModes:
+    - ReadWriteOncePod
+  volumeMode: Filesystem
+  storageClassName: ebs-sc
+  resources:
+    requests:
+      storage: 4Gi
+---
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: ebs-sc
+provisioner: ebs.csi.aws.com
+volumeBindingMode: WaitForFirstConsumer
+parameters:
+  csi.storage.k8s.io/fstype: ext4
+
+```
+Let’s assume that changes have been made in aws-ebs-csi-driver node driver.
+
+**Node publish volume**
+
+1. In the node csi driver, the NodePublishVolume API invokes: `kata-runtime fs add-blk-mount-info “{device: /dev/sdf,fstype: ext4}”`.
```

Comment:
This approach would require the CSI driver to exec and invoke the `kara-runtime` binary and also need to mount the `/run/vc/sbs` sandbox path inside of the CSI driver container correct? I guess if we were to have the CSI driver directly talk to the shim via UDS at `shim-monitor.sock` then it would need to know the sandbox ID? Do we have an approach where we can still directly invoke the shim APIs without needing to do exec and call `kata-runtime`?

Reply:
CSI driver needs to invoke kata-runtime binary, but it doesn't need to mount `/run/vc/sbs` path.

Reply:
`kata-runtime` binary calls the shimclient correct according to: https://github.com/kata-containers/kata-containers/pull/3406/files#diff-48ea539a2381b43645d5e2ae42a582d935385617661359fbabc88affd6e76746R21

so if we deploy the CSI driver as a DaemonSet in K8s, wouldn't the CSI driver need to have something like
```
- hostPath:
     path: /run/vc/sbs
     type: Directory
   name: runtime-monitor
```

in the pod spec in order to connect to the `shim-monitor.sock`?

Reply:
@yibozhuang That's correct. The socket needs to be visible inside the csi node driver container.